### PR TITLE
Ssh schema factory

### PR DESCRIPTION
**Pull request by hensh2ss**
Motivation:

- Refactor the schema directory so that it faciliates extensiblity to sensor update and algorithm updates
- Add a components module to house Component Factories and publisher (pub maps objects to topics and mqtt/kafka publishers)
- Create more unit tests to check the functionality

To Do:
- Check the validity of the thrift conversions in encoding (especially float values)
- Check funtionality with IoT components
- Do testing with sudo real data to check functionality
- complete implementation of Sensor Factory
- Start and complete implmentation of AlgorithmFactory
- Build out integration with publishers

@gwincr11 Please take a look at this (especially the components section and the restructuring of the schemas) and let me know what you think.  In the mean time I am going to work on the precision issue we discussed previously.  Also note that this will require some mods to your serverless to integrate

### CODE REVIEW

**Comment by gwincr11** [comment]
@hensh2ss can you rebase this into a single commit?
git rebase -i HEAD~6

**Comment by gwincr11** [comment]
then force push
git push origin ssh_schema_factory -f

**Comment by hensh2ss** [comment]
@gwincr11 done

File: baa_messages/components/sensor.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import baa_messages.messages.sensor.ttypes as st
+
+
+class SensorFactory:
+    _sensor_module = {}
+
+    def __init__(self):
+        package = baa_messages.messages.sensor
+        prefix = package.__name__ + "."
+        for importer, modname, ispkg in pkgutil.iter_modules(package.__path__, prefix):
```
Comment (gwincr11) [inline baa_messages/components/sensor.py]: Maybe move this into a function, out of the init, to keep the init method clean. The function name will also help describe what this is doing as it is not super clear without some context.

File: baa_messages/components/sensor.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        if "context" in kwargs.keys():
+            ctx = kwargs["context"]
+        else:
+            ctx = BAAContext()
+
+        if "reading" in kwargs.keys():
+            reading = kwargs["reading"]
+        else:
+            exec ("from {0}.ttypes import Reading".format(self._sensor_module[sensor_type]))
+            reading = Reading()
```
Comment (gwincr11) [inline baa_messages/components/sensor.py]: Will people want to be able to pass in the initlization params for Reading, Context, etc...?

File: baa_messages/components/sensor.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+        return st.SensorReading(ctx, sr)
+
+    def create_sensor_setting(self, sensor_type, **kwargs):
+        if sensor_type not in self._sensor_module:
+            raise ValueError("Invalid sensor_type: {0} in {1}".format(sensor_type,
+                                                                      "SensorFactory.create_sensor_reading"))
+
+        # Checking kwargs for context or reading
+        if "context" in kwargs.keys():
```
Comment (gwincr11) [inline baa_messages/components/sensor.py]: This context check is duplicated can you Dry it up.

File: baa_messages/components/sensor.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        else:
+            ctx = BAAContext()
+
+        if "setting" in kwargs.keys():
+            setting = kwargs["setting"]
+        else:
+            exec ("from {0}.ttypes import Setting".format(self._sensor_module[sensor_type]))
+            setting = Setting()
+
+        for k, v in kwargs.iteritems():
```
Comment (gwincr11) [inline baa_messages/components/sensor.py]: This iterator is duplicated, you can dry it up

File: baa_messages/components/sensor.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                if hasattr(i, k):
+                    setattr(i, k, v)
+
+        sensor_setting = st.Setting()
+        setattr(sensor_setting, sensor_type, setting)
+
+        return st.SensorSetting(ctx, sensor_setting)
+
+    def create_sensor_report(self,**kwargs):
+        if "context" in kwargs.keys():
```
Comment (gwincr11) [inline baa_messages/components/sensor.py]: Here is the context check again.

File: test/test_sensor.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+# from baa_messages.messages.sensor.gps.ttypes import Reading as GPSReading
+# import baa_messages.codec as bc
+# import time
+# from baa_messages.messages.core.ttypes import BAAContext
+# from baa_messages.messages.sensor.ttypes import SensorReading
+# import baa_messages
+
+
+class TestSensor(unittest.TestCase):
+    def test_individual_constructors(self):
```
Comment (gwincr11) [inline test/test_sensor.py]: There should be a test like this that passes in a non existent name so that we know the error checking works.

File: test/test_sensor.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        print 'Constructing Sensor Reading'
+        sr = st.SensorReading(ctx,st.Reading(gr))
+        print 'Constructing Sensor Setting'
+        ss = st.SensorSetting(ctx,st.Setting(gs))
+
+        print 'Constructing Sensor Report'
+        my_rep = st.SensorReport(ctx,[sr,sr],[ss])
+
+        print 'Serializing Sensor Report into TBinary'
+        payload_bytes = bc.encode_object(my_rep,bc.Encoding.TSIMPLEJSON)
```
Comment (gwincr11) [inline test/test_sensor.py]: This test is not actually asserting anything? Did you forget to add your test?

File: test/test_sensor.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+    def test_sensor_factory(self):
+        print 'Testing components.sensor.SensorFactory...'
+
+        from baa_messages.components.sensor import SensorFactory
+        sf = SensorFactory()
+        gr = sf.create_sensor_reading("gps")
+        self.assertRaises(ValueError,sf.create_sensor_reading,"time")
```
Comment (gwincr11) [inline test/test_sensor.py]: It is easier to see what tests are passing or failing when 1 test = 1 piece of functionality. I would move this into a separate test.

File: test/test_sensor.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        self.assertEqual(gr2.reading.gps.longitude,lng)
+
+
+        temp_reading = sf.create_sensor_reading("temperature",temperature=98.7)
+        gamma_reading = sf.create_sensor_setting("gamma",high_voltage=1000)
+
+        gt = sf.create_sensor_setting("gps",timestamp=time.time())
+
+        sr = sf.create_sensor_report(readings=[gr,gr2],settings=[gt])
+        payload_bytes = bc.encode_object(sr,bc.Encoding.TBINARY)
```
Comment (gwincr11) [inline test/test_sensor.py]: Looks like you are setting up to test encoding. I think we should have a seperate test file for encodings.

File: test/test_sensor.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+
+if __name__=="__main__":
+    unittest.main()
```
Comment (gwincr11) [inline test/test_sensor.py]: Give this a try: http://stackoverflow.com/a/27293806/319342

**Comment by gwincr11** [comment]
Overall this is good progress and IMO you are moving in the right direction. Here are my thoughts:

Notes:
1. I think it would be smart to add a "model" test for each thrift schema so we can test that it accepts the type of data we expect.
2. Perhaps we need a better description for parent_id (BAA_Context) and receiver (BAA_Messages) fields. I like the graph viz docs is there a way to add descriptions? Is there a better name?
3. Is Golf algorithm a particular algorithm, or am I missing a piece of nomenclature?
4. What will the IPC Message be?
5. Is the BAA_Context going to hold the setting time for SensorSetting? 

I need to better understand the sensor unit id, sensor id and timestamp schema locations as these will be the keys I work with. Am I correct in thinking I will always be pulling that from BAA_Context? If so I think we should validate that this is attached to any message going through the publisher and all those fields are set before allowing it to publish.

Along those lines do we need to validate that a minimum number of fields are set before we allow them to send? Does Thrift have a required fields/validate concept?

File: baa_messages/components/sensor.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
+import pkgutil
+import baa_messages.messages.sensor
+from baa_messages.messages.core.ttypes import BAAContext
+import baa_messages.messages.sensor.ttypes as st
+
+
+class SensorFactory:
```
Comment (gwincr11) [inline baa_messages/components/sensor.py]: I am not sure how all the pieces will endup coming together, but some of these functions maybe better suited in their own classes and being called out to so that an algorithm factory, sensor unit factory etc... can also use them.

File: baa_messages/components/sensor.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+
+
+# from baa_messages.publisher.mqtt import Mqtt
+# from baa_messages.publisher.publisher import Broadcaster
+# from baa_messages.codec import create_network_message
+# import time
+#
+#
+# class SensorPublisher(Broadcaster):
```
Comment (gwincr11) [inline baa_messages/components/sensor.py]: I would start by implementing sensor unit and see how you want to work with individual sensor's by writing psuedo code then trying to implement it.

File: baa_messages/components/sensor.py (hunk 1 above)
Comment (hensh2ss) [inline baa_messages/components/sensor.py]: Good thought I will make this update.

File: baa_messages/components/sensor.py (hunk 2 above)
Comment (hensh2ss) [inline baa_messages/components/sensor.py]: I think so, or at least they should have the ability to then they can choose.

File: baa_messages/components/sensor.py (hunk 3 above)
Comment (hensh2ss) [inline baa_messages/components/sensor.py]: I am not sure what you mean here.  Line 39-41 checks the kwargs for keys specific to a BAAContext or Reading and then sets the value for the reading.

File: test/test_sensor.py (hunk 6 above)
Comment (hensh2ss) [inline test/test_sensor.py]: done

File: test/test_sensor.py (hunk 7 above)
Comment (hensh2ss) [inline test/test_sensor.py]: fixed

File: test/test_sensor.py (hunk 9 above)
Comment (hensh2ss) [inline test/test_sensor.py]: done

**Comment by hensh2ss** [comment]
@gwincr11 check out the updates...more to come but covered alot of the comments

**Comment by gwincr11** [comment]
Hey @hensh2ss 

This pull request is getting hard to follow. It seems like your rebase did not work, if it worked there should only be 1 commit in the pull request.

I am also not sure if you didn't commit some of the changes you made, but it seems like you maybe didn't push up some changes you made if you are saying a lot of the comments I made have been addressed? Am I missing something?

Cory

**Comment by hensh2ss** [comment]
Ina meeting today, well try to look tomorrow

**Comment by hensh2ss** [comment]
@gwincr11 Yeah the rebase didn't work previously (I am a nube to rebase)  I believe I have this resolved now...take a look and let me know what you think

File: baa_messages/publisher/mqtt.py [hunk 13]
```diff
@@ -9,5 +10,7 @@ def __init__(self, topic="", sensor_id="", credentials_dir=False):
     def destroy(self):
         self.mqClient.disconnect()
 
-    def publish(self, msg):
+    def publish(self, msg, topic=None):
+        if topic is not None:
+            self.set_topic(topic)
```
Comment (gwincr11) [inline baa_messages/publisher/mqtt.py]: This is a nice improvement.

Comment (gwincr11) [inline baa_messages/publisher/mqtt.py]: Is there a way to put this in the abstract class and call it as a super method?

**Comment by gwincr11** [comment]
Cool, looks like things are moving well. Are their any breaking changes? If not lets merge in and start from scratch with the next set of features.
